ctwdtw/essential-feed-case-study
# testing view controller with view model

- test `FeedViewController` can init
- use factory method `makeSUT` to extract the init logic
- move `UIImageTransformer` to a separate file
- rename test double
- show loading indicator on view presented
- hide loading indicator on loading finished
- add helper function to enhance test case readability
- use `XCTUnwrap`, `#filePath`, `#line` to enhance readability of failing test message
- instead of testing `FeedViewController` Integratively, testing the collaboration between `FeedViewController` and `FeedViewModel`
- show loading indicator when user initiate a loading action
- add more assertions to verify `loadFeed ` is called
- add DSL to hide implementation detail from tests
- adjust vertical spacing to make test case more readable
- instead of calling observer closure directly, manipulating state on `FeedViewModel` seems to be a more sensible way to implement test cases
- render `FeedImageViewMode`s
- use implicitly optional for  `FeedViewController`'s dependency on `FeedViewModel`
- trigger image data loading on feed image view is visible
- cancel image data loading on feed image view not visible anymore
- change access control of `isLoadingImage` to internal
- show feed image view loading indicator (shimming) while loading image
- feed image view renders image loaded from url
- add assertion message to enhance test case readability
- retry button is visible when `shouldRetry` state of `FeedImageViewModel` become true
- use `ImageLoadingState` to group unorganized states such as `shouldRetry`, `image`, `isLoadingImage` ... etc. together
- add `MARK:-` for `FeedImageViewModel
- update test case name and test case message
- feed image retry action retries image load
- feed image view preload image when near visible
- feed image view cancel image preloading when not near visible anymore
